### PR TITLE
#93 - implemented ABI encoding

### PR DESCRIPTION
**Pull request by biboran**
To solve #93 I added ABI encoding of the Solidity types.

There is a constraint with enclosed tuples (fixed length arrays) which requires users to flatmap tuples to the root tuple.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#95](https://github.com/blockstoreapp/web3swift/issues/95) is now in scope, role is `REV`

**Comment by biboran** [comment]
> What is `tails`

@rockfridrich I took the idea of heads-tails encoding from [abi specification](https://solidity.readthedocs.io/en/develop/abi-spec.html#formal-specification-of-the-encoding) tuple section.

**Comment by rockfridrich** [comment]
@biboran does user really need to know about this during parameter encoding? Should not it be hidden inside?

File: Example/Tests/ABI/ABIStringTests.swift [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+                )
+            ),
+            description: "ABI string is expected to encode length to tails correctly"
+        )
+    }
+
+    func testEncodesTailsCorrectly() {
+        expect{
+            try ConcatenatedBytes(
+                bytes: ABIString(
```
Comment (rockfridrich) [inline Example/Tests/ABI/ABIStringTests.swift]: @biboran and what about function signature parameter that takes only first 8 (without 0x) letters?

Comment (biboran) [reply Example/Tests/ABI/ABIStringTests.swift]: @rockfridrich that is not what is tested here. Content of the string here is not important.

Comment (biboran) [reply Example/Tests/ABI/ABIStringTests.swift]: @rockfridrich you can see a proper signature encoding in the `EncodedABIFunctionTests`

**Comment by biboran** [comment]
> @biboran does user really need to know about this during parameter encoding? Should not it be hidden inside?

@rockfridrich for user only `EncodedABIFunction` interface is required to be able to interact with the node.

File: Web3Swift/ABI/ABIBoolean.swift [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+                        return Data(
+                            bytes: [0x01]
+                        )
+                    } else {
+                        return Data(
+                            bytes: [0x00]
+                        )
+                    }
+                },
+                padding: 32
```
[the remark this answers is not in the document]
Comment (biboran) [reply Web3Swift/ABI/ABIBoolean.swift]: @rockfridrich I can wrap 32 into a couple objects that will have 32 hardcoded but I don't see any point. ABI business logic is pretty atomic.

File: Web3Swift/ABI/EncodedABIFunction.swift [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    Encoded function as `Data`
+
+    - throws:
+    `DescribedError` if something went wrong.
+    */
+    public func value() throws -> Data {
+        return try ConcatenatedBytes(
+            bytes: [
+                FixedLengthBytes(
+                    origin: FirstBytes(
```
[the remark this answers is not in the document]
Comment (biboran) [reply Web3Swift/ABI/EncodedABIFunction.swift]: @rockfridrich `FixedLengthBytes` makes sure there are only `n` bytes while `FirstBytes` takes first bytes of a bytes sequence up to `n`. So `FirstBytes` can actually represent 0 to n bytes.

**Comment by 0crat** [comment]
The job [#95](https://github.com/blockstoreapp/web3swift/issues/95) is now out of scope

**Pull request merged**